File: src/camera.nim
import math

import ray
import vec
import util


type
  Camera* = object
    lower_left_corner*: Vector3d
    horizontal*: Vector3d
    vertical*: Vector3d
    origin*: Vector3d
    lens_radius*: float
    u, v, w: Vector3d


proc random_in_unit_disk: Vector3d =
  var p: Vector3d
  doWhile dot(p, p) >= 1.0:
    p = vector3d(drand48(), drand48(), 0.0) * 2 - vector3d(1.0, 1, 0)
  return p


proc newCamera*(lookfrom, lookat, vup: Vector3d, vfov, aspect, aperture, focus_dist: float): Camera =
  let
    theta = vfov * (PI / 180)
    half_height = tan(theta / 2)
    half_width = aspect * half_height

  result.origin = lookfrom
  result.lens_radius = aperture / 2
  result.w = unit(lookfrom - lookat)
  result.u = unit(cross(vup, result.w))
  result.v = cross(result.w, result.u)
  result.lower_left_corner = result.origin - (result.u * half_width * focus_dist) -
                            (result.v * half_height * focus_dist) -
                            (result.w * focus_dist)
  result.horizontal = result.u * 2 * half_width * focus_dist
  result.vertical = result.v * 2 * half_height * focus_dist

proc ray*(camera: Camera, s, t: float): Ray =
  let
    rd = random_in_unit_disk() * camera.lens_radius
    offset = camera.u * rd.x + camera.v * rd.y
    direction = camera.lower_left_corner + camera.horizontal * s + camera.vertical * t - camera.origin - offset
  result = ray(camera.origin + offset, direction)

File: src/camera.py
import math
import random

from vectors import Vector
from ray import Ray


def random_in_unit_disk():
    p = Vector(random.random(), random.random(), 0.0) * 2 - Vector(1.0, 1.0)
    while p.dot(p) >= 1.0:
        p = Vector(random.random(), random.random(), 0.0) * 2 - Vector(1.0, 1.0)
    return p


class Camera(object):

    def __init__(self, lookfrom, lookat, vup, vfov, aspect, aperture, focus_dist):
        theta = vfov * (math.pi / 180)
        half_height = math.tan(theta / 2)
        half_width = aspect * half_height

        self.origin = lookfrom
        self.lens_radius = aperture / 2
        self.w = Vector.unit(lookfrom - lookat)
        self.u = Vector.unit(vup.cross(self.w))
        self.v = self.w.cross(self.u)
        self.lower_left_corner = self.origin - (self.u * half_width * focus_dist) - (self.v * half_height * focus_dist) - (self.w * focus_dist)
        self.horizontal = self.u * 2 * half_width * focus_dist
        self.vertical = self.v * 2 * half_height * focus_dist

    def ray(self, s, t):
        rd = random_in_unit_disk() * self.lens_radius
        offset = self.u * rd.x + self.v * rd.y
        direction = self.lower_left_corner + self.horizontal * s + self.vertical * t - self.origin - offset

        return Ray(self.origin + offset, direction)
